src/drivers/intel/dmic/dmic_nhlt.c

```diff
@@ -0,0 +1,418 @@
+
+		} else {
+			dmic->dai_channels = 1;
+			dmic->enable[0] = 0x0; /* PDM0 none */
+			mic_swap = MIC_CONTROL_PDM_CLK_EDGE_GET(pdm_cfg[1]->mic_control);
+			dmic->enable[1] = mic_swap ? 0x2 : 0x1; /* PDM1 MIC B or MIC A */
+		}
+		break;
+	case 2:
+		if (fir_stereo) {
+			dmic->dai_channels = 4;
+			dmic->enable[0] = 0x3; /* PDM0 MIC A and B */
+			dmic->enable[1] = 0x3;	/* PDM1 MIC A and B */
+		} else {
+			dai_info(dai, "nhlt_dmic_dai_params_get(): Illegal 4ch configuration");
+			return -EINVAL;
+		}
+		break;
+	default:
+		dai_info(dai, "nhlt_dmic_dai_params_get(): Illegal OF bit field");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int dmic_set_config_nhlt(struct dai *dai, void *spec_config)
+{
+	struct dmic_pdata *dmic = dai_get_drvdata(dai);
+	struct nhlt_pdm_ctrl_cfg *pdm_cfg[DMIC_HW_CONTROLLERS_MAX];
+	struct nhlt_pdm_ctrl_fir_cfg *fir_cfg_a[DMIC_HW_CONTROLLERS_MAX];
+	struct nhlt_pdm_ctrl_fir_cfg *fir_cfg_b[DMIC_HW_CONTROLLERS_MAX];
+	struct nhlt_pdm_fir_coeffs *fir_a[DMIC_HW_CONTROLLERS_MAX] = {NULL};
+	struct nhlt_pdm_fir_coeffs *fir_b[DMIC_HW_CONTROLLERS_MAX];
+	uint32_t out_control[DMIC_HW_FIFOS_MAX];
+	uint32_t channel_ctrl_mask;
+	uint32_t pdm_ctrl_mask;
+	uint32_t val;
+	uint32_t fir_control;
+	const uint8_t *p = spec_config;
+	int num_fifos;
+	int num_pdm;
+	int fir_length_a;
+	int fir_length_b;
+	int n;
+	int i;
+	int p_mcic = 0;
+	int p_mfira = 0;
+	int p_mfirb = 0;
+	int p_clkdiv = 0;
+	int clkdiv;
+	int comb_count;
+	int fir_decimation, fir_shift, fir_length;
+	int bf1, bf2, bf3, bf4, bf5, bf6, bf7, bf8;
+	int ret;
+
+	/* p = &nhlt_dmic_cfg[0]; */ /* Temporary blob for tests, will be replaced by IPC */
+
+	/* Skip not used headers */
+	p += sizeof(struct nhlt_dmic_gateway_attributes);
+	p += sizeof(struct nhlt_dmic_ts_group);
+	p += sizeof(struct nhlt_dmic_clock_on_delay);
+
+	/* Channel_ctlr_mask bits indicate the FIFOs enabled*/
+	channel_ctrl_mask = ((struct nhlt_dmic_channel_ctrl_mask *)p)->channel_ctrl_mask;
+	num_fifos = count_bits_set(channel_ctrl_mask);
+	p += sizeof(struct nhlt_dmic_channel_ctrl_mask);
+	dai_info(dai, "dmic_set_config_nhlt(): channel_ctrl_mask = %d", channel_ctrl_mask);
+
+	/* Get OUTCONTROLx configuration */
+	if (num_fifos < 1 || num_fifos > DMIC_HW_FIFOS_MAX) {
+		dai_err(dai, "dmic_set_config_nhlt(): illegal number of FIFOs %d", num_fifos);
+		return -EINVAL;
+	}
+
+	for (n = 0; n < num_fifos; n++) {
+		val = *(uint32_t *)p;
+		out_control[n] = val;
+		bf1 = OUTCONTROL0_TIE_GET(val);
+		bf2 = OUTCONTROL0_TIE_GET(val);
+		bf3 = OUTCONTROL0_FINIT_GET(val);
+		bf4 = OUTCONTROL0_FCI_GET(val);
+		bf5 = OUTCONTROL0_BFTH_GET(val);
+		bf6 = OUTCONTROL0_OF_GET(val);
+		bf7 = OUTCONTROL0_IPM_GET(val);
+		bf8 = OUTCONTROL0_TH_GET(val);
+		dai_info(dai, "dmic_set_config_nhlt(): OUTCONTROL%d = %08x",
+			 n, out_control[n]);
+		dai_info(dai, "  tie=%d, sip=%d, finit=%d, fci=%d", bf1, bf2, bf3, bf4);
+		dai_info(dai, "  bfth=%d, of=%d, ipm=%d, th=%d", bf5, bf6, bf7, bf8);
```

Comment:
do we need this information in traces? I understand it's useful to you now for debugging, maybe make it `dai_dbg()` for upstream?

Reply:
I've cleaned  up the traces. Most dai_info()s are now dai_dbg().